lnrpc/routerrpc/router_server.go

```diff
@@ -1043,6 +1048,149 @@ func (s *Server) SendOnion(_ context.Context,
+	// to update the Switch such that it doesn't fall over with a nil
+	// error decryptor. In this scenario, we should defer all error handling
+	// and return raw error blobs to the caller.
+	log.Debugf("Looking up status of onion payment for HTLC hash %x and "+
+		"attempt_id=%d", req.PaymentHash, req.AttemptId)
+
+	// Attempt to build the error decryptor with the provided session key
+	// and hop public keys.
+	errorDecryptor, err := buildErrorDecryptor(
+		req.SessionKey, req.HopPubkeys,
+	)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"error building decryptor: %v", err)
+	}
+
+	if errorDecryptor == nil {
+		log.Debug("Unable to build error decrypter with information " +
+			"provided. Will defer error handling to caller")
+	}
+
+	// Query the switch for the result of the payment attempt via onion.
+	resultChan, err := s.cfg.HtlcDispatcher.GetAttemptResult(
+		req.AttemptId, lntypes.Hash(req.PaymentHash), errorDecryptor,
+	)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal,
+			"failed locate payment attempt: %v", err)
+	}
+
+	// The switch knows about this payment, we'll wait for a result to be
+	// available.
+	var (
+		result *htlcswitch.PaymentResult
+		ok     bool
+	)
+
+	select {
+	case result, ok = <-resultChan:
+		if !ok {
+			// NOTE(calvin): Can RPC clients see this error type or
+			// will they need to string match?
+			return nil, status.Errorf(codes.Internal,
+				"failed locate payment attempt: %v",
+				htlcswitch.ErrSwitchExiting)
+		}
+
+	case <-ctx.Done():
+		return nil, nil
+	}
+
+	// In case of a payment failure, fail the attempt with the control
+	// tower and return.
+	if result.Error != nil {
+		log.Errorf("Payment via onion failed for hash %x",
+			req.PaymentHash)
+
+		// TODO(calvin): Check error handling to see if we need to
+		// include a code.
+		message, _ := htlcswitch.TranslateErrorForRPC(result.Error)
+
+		return &TrackOnionResponse{
+			Success:      false,
+			ErrorMessage: message,
+		}, nil
+	}
+
+	// In case we don't process the error, we'll return the encrypted
+	// error blob for handling by the caller.
+	if result.EncryptedError != nil {
+		log.Errorf("Payment via onion failed for hash %x",
+			req.PaymentHash)
+
+		// NOTE(calvin): gRPC will not return a response object if the
+		// error is non-nil. So if we want to return the encrypted
+		// error blob to the caller, then we need to return a nil error.
+		return &TrackOnionResponse{
+			Success:        false,
+			EncryptedError: result.EncryptedError,
+		}, nil
+	}
+
+	return &TrackOnionResponse{
+		Success:  true,
+		Preimage: result.Preimage[:],
+	}, nil
+}
+
+func buildErrorDecryptor(sessionKeyBytes []byte,
+	hopPubkeys [][]byte) (htlcswitch.ErrorDecrypter, error) {
+
+	if len(sessionKeyBytes) == 0 || len(hopPubkeys) == 0 {
+		return nil, nil
+	}
+
+	// TODO(calvin): Validate that the session key makes a valid private
+	// key? This is untrusted input received via RPC.
```

Comment:
+1
I posted a comment for `Server.BuildOnion` method. I think it makes sense to reuse the same validation code here and there.